File: scripts/ingest-sample.js
#!/usr/bin/env node
/**
 * Hello-world ingest: upsert one corpus requirement into SurrealDB and read it back.
 *
 *   npm run db:ingest-sample
 */

import { readFileSync } from "node:fs";
import { execFileSync } from "node:child_process";
import { RecordId } from "surrealdb";
import { connectSurreal } from "../src/db/connect.js";

function parseYaml(text) {
  return JSON.parse(
    execFileSync(
      "python3",
      [
        "-c",
        "import sys,yaml,json; print(json.dumps(yaml.safe_load(sys.stdin.read())))",
      ],
      { input: text, encoding: "utf8" },
    ),
  );
}

const path =
  process.argv[2] ||
  "frameworks/annex-sl-core/1.0/requirements/4.1-context.yaml";

try {
  const doc = parseYaml(readFileSync(path, "utf8"));
  if (!doc?.id) throw new Error(`YAML at ${path} is missing required field id`);

  const db = await connectSurreal();
  try {
    const key = doc.id;
    const rid = new RecordId("requirement", key);
    const content = {
      ...doc,
      source_path: path,
      ingested_at: new Date().toISOString(),
    };

    const written = await db.upsert(rid).content(content);
    const selected = await db.query(
      `SELECT id, title, framework, ref, source_path, domains FROM type::record("requirement", $key);`,
      { key },
    );

    console.log("ingest-sample OK");
    console.log(JSON.stringify({ key, written, selected }, null, 2));
  } finally {
    await db.close();
  }
  process.exit(0);
} catch (err) {
  console.error("ingest-sample FAILED");
  console.error(err?.stack || err?.message || err);
  process.exit(1);
}

File: lib/isomer/db/ingest_sample.ex
defmodule Isomer.Db.IngestSample do
  @moduledoc "Hello-world ingest: upsert one corpus requirement and read it back."

  alias Isomer.Db.{Connect, Params}
  alias Isomer.YAML

  @default_path "frameworks/annex-sl-core/1.0/requirements/4.1-context.yaml"

  def run!(path \\ @default_path) do
    abs = Path.expand(path, Isomer.root())
    doc = YAML.read!(abs)

    unless is_binary(doc["id"]) and doc["id"] != "" do
      raise ArgumentError, "YAML at #{path} is missing required field id"
    end

    key = doc["id"]
    db = Connect.connect!()

    try do
      fields = Map.drop(doc, ["id"])

      content =
        fields
        |> Map.put("corpus_id", key)
        |> Map.put("content_source", Params.content_source())
        |> Map.put("source_path", path)
        |> Map.put("ingested_at", DateTime.utc_now() |> DateTime.to_iso8601())

      rid = SurrealDB.RecordId.new("requirement", key)

      {:ok, written} = SurrealDB.upsert(db, rid, content)

      {:ok, selected} =
        SurrealDB.query(
          db,
          ~s|SELECT id, title, framework, ref, source_path, domains FROM type::record("requirement", $key);|,
          %{"key" => key}
        )

      %{key: key, written: written, selected: selected}
    after
      SurrealDB.close(db)
    end
  end
end
